FastExcel/DefinedName.cs

```diff
@@ -0,0 +1,201 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml.Linq;
+
+namespace FastExcel
+{
+    partial class FastExcel
+    {
+        /// <summary>
+        /// Dictionary of defined names
+        /// </summary>
+        internal IReadOnlyDictionary<string, DefinedName> DefinedNames { get; private set; }
+
+        private void loadDefinedNames()
+        {
+            XDocument document;
+            try
+            {
+                document = XDocument.Load(Archive.GetEntry("xl/workbook.xml").Open());
+            }
+            catch (Exception exception)
+            {
+                throw new DefinedNameLoadException("Unable to open stream to read internal workbook.xml file", exception);
+            }
+            if (document == null)
+            {
+                throw new DefinedNameLoadException("Unable to load workbook.xml from open stream. Probable corrupt file.");
+            }
+
+            var definedNames = new Dictionary<string, DefinedName>();
+
+            foreach (var e in (from d2 in document.Descendants().Where(dx => dx.Name.LocalName == "definedNames").Descendants()
+                               select d2))
+            {
+                var currentDefinedName = new DefinedName(e);
+                definedNames.Add(currentDefinedName.Key, currentDefinedName);
+            }
+            DefinedNames = definedNames;
+        }
+
+        /// <summary>
+        /// Retrieves ranges of cells by their defined name
+        /// </summary>
+        /// <param name="definedName">Defined Name</param>
+        /// <param name="worksheetIndex">If scoped to a sheet, the worksheetIndex</param>
+        /// <returns>List of cells encapsulated in another list representing seperate ranges</returns>
+        public IEnumerable<IEnumerable<Cell>> GetCellRangesByDefinedName(string definedName, int? worksheetIndex = null)
+        {
+            List<List<Cell>> result = new List<List<Cell>>();
+
+            string key = (!worksheetIndex.HasValue ? definedName : definedName + ":" + worksheetIndex);
+
+            if (!DefinedNames.ContainsKey(key))
+                return result;
+
+            string[] references = DefinedNames[key].Reference.Split(',');
+
+            foreach(string reference in references)
+            {
+                // If A) is a formula or B) does not contain ! or $ then this reference is not supported
+                if (Regex.IsMatch(reference, @"[\(\)\*\+\-\/]") || !Regex.IsMatch(reference, @"[!$]"))
```

Comment:
This should check ! and $ is in the string. Second regular expression should be `!.*\$`